### DISCUSSION TITLE
How to install providers correctly?

### DISCUSSION THREAD

**Discussion by BokDan93** [xtekky/gpt4free#1816] [Q&A]
I have a function where I alternately access different models. I added all the providers that I found in the models file for the models I needed. What if I don’t specify providers, then on what basis do we use providers to get an answer? Are they taken randomly?

What is the difference between these functions?
client = G4fClient(provider=RetryProvider([You, ChatgptNext, OpenaiChat, Bing, Liaobots]))
or
client = G4fClient()

Here is my code:

models = ["gpt-4-turbo", "gpt-3.5-turbo", "gpt-4", "gpt-3.5-turbo"]
models = models * 3

    for attempt in range(1, 13):
        client = G4fClient(provider=RetryProvider([You, ChatgptNext, OpenaiChat, Bing, Liaobots]))
        loop = asyncio.get_event_loop()
        try:
            model = models[attempt - 1]
            response = await loop.run_in_executor(None, lambda: client.chat.completions.create(
                model=model,
                messages=[{"role": "user", "content": full_request}],
                web_search=False
            ))
            response_str = response.choices[0].message.content

**Comment by hlohaus** [comment]
Hey,

when you don't choose a provider, it uses the "best_provider" from the models file. We're using the RetryProvider to pick a provider from a list of "best" providers.

**Comment by hlohaus** [comment]
You can also use the AsyncClient. Then you don't need to loop the call. It's already asynchronous. Check out the docs here: [g4f docs](https://github.com/xtekky/gpt4free/blob/main/docs/async_client.md)

**Comment by BokDan93** [comment]
Thank you, I did as you said, I received an answer, but now I have an error:
Python\Python310\lib\site-packages\curl_cffi\aio.py:205: UserWarning: Curlm alread closed! quitting from process_data
  warnings.warn("Curlm alread closed! quitting from process_data")
